File: donghwa/arxiv/palindrome/2697_lexi_smallest_palindrome.c
/*
You are given a string s consisting of lowercase English letters, and you are allowed to perform operations on it. In one operation, you can replace a character in s with another lowercase English letter.

Your task is to make s a palindrome with the minimum number of operations possible. If there are multiple palindromes that can be made using the minimum number of operations, make the lexicographically smallest one.

A string a is lexicographically smaller than a string b (of the same length) if in the first position where a and b differ, string a has a letter that appears earlier in the alphabet than the corresponding letter in b.

Return the resulting palindrome string.

 

Example 1:

Input: s = "egcfe"
Output: "efcfe"
Explanation: The minimum number of operations to make "egcfe" a palindrome is 1, and the lexicographically smallest palindrome string we can get by modifying one character is "efcfe", by changing 'g'.
Example 2:

Input: s = "abcd"
Output: "abba"
Explanation: The minimum number of operations to make "abcd" a palindrome is 2, and the lexicographically smallest palindrome string we can get by modifying two characters is "abba".
Example 3:

Input: s = "seven"
Output: "neven"
Explanation: The minimum number of operations to make "seven" a palindrome is 1, and the lexicographically smallest palindrome string we can get by modifying one character is "neven".

*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int min(int a, int b) {
    return a < b ? a : b;
}

char* makePalindrome(char* s) {
    int len = strlen(s);
    char* result = strdup(s);
    int operations = 0;
    int left = 0, right = len - 1;

    while (left <= right) {
        if (result[left] != result[right]) {
            result[left] = result[right] = min(result[left], result[right]);
            operations++;
        }
        left++;
        right--;
    }

    // If there are still more than one operations needed, repeat the process
    if (operations > 1) {
        return makePalindrome(result);
    }

    return result;
}

int main() {
    char s1[] = "egcfe";
    char* result1 = makePalindrome(s1);
    printf("Example 1 Output: %s\n", result1);  // Output: "efcfe"
    free(result1);

    char s2[] = "abcd";
    char* result2 = makePalindrome(s2);
    printf("Example 2 Output: %s\n", result2);  // Output: "abba"
    free(result2);

    char s3[] = "seven";
    char* result3 = makePalindrome(s3);
    printf("Example 3 Output: %s\n", result3);  // Output: "neven"
    free(result3);

    return 0;
}

File: donghwa/arxiv/palindrome/2697_lexi_smallest_palindrome.py
def makePalindrome(s):
    def isPalindrome(s):
        return s == s[::-1]

    s = list(s)
    operations = 0
    left, right = 0, len(s) - 1

    while left <= right:
        if s[left] != s[right]:
            s[left] = min(s[left], s[right])
            s[right] = s[left]
            operations += 1
        left += 1
        right -= 1

    # If there are still more than one operations needed, repeat the process
    if operations > 1:
        return makePalindrome("".join(s))

    return "".join(s)

# Test cases
s1 = "egcfe"
print("Example 1 Output:", makePalindrome(s1))  # Output: "efcfe"

s2 = "abcd"
print("Example 2 Output:", makePalindrome(s2))  # Output: "abba"

s3 = "seven"
print("Example 3 Output:", makePalindrome(s3))  # Output: "neven"
